sycl/include/CL/sycl/detail/config.hpp

```diff
@@ -0,0 +1,75 @@
+//==---------------- config.hpp - SYCL context ------------------*- C++-*---==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <cstring>
+#include <fstream>
+#include <iostream>
+#include <limits>
+#include <sstream>
+
+namespace cl {
+namespace sycl {
+namespace detail {
+
+// Enum of config IDs for accessing other arrays
+enum ConfigID {
+  START = 0,
+#define CONFIG(name, ...) name,
+#include "config.def"
+#undef CONFIG
+  END
+};
+
+// Consider strings starting with __ as unset
+constexpr const char *getStrOrNullptr(const char *Str) {
+  return (Str[0] == '_' && Str[1] == '_') ? nullptr : Str;
+}
+
+template <ConfigID Config> class SYCLConfigBase;
+
+#define CONFIG(Name, MaxSize, CompileTimeDef)                                  \
+  template <> class SYCLConfigBase<Name> {                                     \
+  public:                                                                      \
+    /*Preallocated storage for config value which is extracted from a config   \
+     * file*/                                                                  \
+    static char MStorage[MaxSize];                                             \
+    /*Points to the storage if config is set in the file, nullptr otherwise*/  \
+    static const char *MValueFromFile;                                         \
+    /*The name of the config*/                                                 \
+    static constexpr const char *const MConfigName = STRINGIFY_LINE(Name);     \
+    /*Points to the value which is set during compilation, nullptr otherwise.  \
+     * Detection of whether a value is set or not is based on checking the     \
```

Comment:
I'm not sure why you want to check whether it was set or not. Why not just have a default value for all configuration options, and specify them in config.def instead of a prefixed name?

Reply:
Ok, do you suggest writing CONFIG in the following way:
have default:
CONFIG(SYCL_PRINT_EXECUTION_GRAPH, 32, "SUPER DEFAULT VALUE")
have no default:
CONFIG(SYCL_PRINT_EXECUTION_GRAPH, 32, nullptr)
?
This is an option, but in this case you are not able to just pass -DSYCL_PRINT_EXECUTION_GRAPH to the compiler, you need to modify sources(.def) instead.

Do you think it's OK?

Reply:
> Ok, do you suggest writing CONFIG in the following way:
> have default:
> CONFIG(SYCL_PRINT_EXECUTION_GRAPH, 32, "SUPER DEFAULT VALUE")
> have no default:
> CONFIG(SYCL_PRINT_EXECUTION_GRAPH, 32, nullptr)

No, I'm not sure how a configuration value may have no default value at all. 

Take SYCL_PRINT_EXECUTION_GRAPH for example and let's suppose I don't have it in the config and the environment variable is set. Am I going to see a graph or not? If not, then the default value is "0".

Reply:
There are cases where we check that some env variable exists and actual content doesn't matter: `if(getenv("IMPROVE_PERFORMANCE_100X"))`
getenv returns nullptr if value is not set which is not equal to "0" value, so config helper mimics this behavior.

Reply:
Discussed offline.
It's weird to have 3 conditions for essentially boolean values (true, false and a mysterious unset), but we don't want to do string comparison every time for performance reasons. Obvious solution is to use enums and integers instead of strings, but this can be implemented later.